sdk/core/core/src/az_http_policy.c

```diff
@@ -61,12 +65,110 @@ AZ_NODISCARD az_result az_http_pipeline_policy_retry(
+  while (true) {
+    AZ_RETURN_IF_FAILED(az_http_response_reset(response));
+    result = az_http_pipeline_nextpolicy(p_policies, hrb, response);
+
+    // Even HTTP 429, or 502 are expected to be AZ_OK, so the failed result is not retriable.
+    if (attempt > max_tries || az_failed(result) || response == NULL
+        || response->builder.length == 0) {
+      return result;
+    }
+
+    int32_t retry_after_msec = -1;
+    {
+      az_http_response_parser response_parser = { 0 };
+      AZ_RETURN_IF_FAILED(az_http_response_parser_init(
+          &response_parser, az_span_builder_result(&response->builder)));
+
+      az_http_response_status_line status_line = { 0 };
+      AZ_RETURN_IF_FAILED(az_http_response_parser_read_status_line(&response_parser, &status_line));
+
+      switch (status_line.status_code) {
+        // Keep retrying on the HTTP response codes below, return immediately otherwise.
+        case 408:
+        case 429:
+        case 500:
+        case 502:
+        case 503:
+        case 504: {
+          // Try to get the value of retry-after header, if there's one.
+          az_pair header = { 0 };
+          while (az_http_response_parser_read_header(&response_parser, &header) == AZ_OK) {
+            if (az_span_is_equal(header.key, AZ_STR("retry-after-ms"))
+                || az_span_is_equal(header.key, AZ_STR("x-ms-retry-after-ms"))) {
+              // The value is in milliseconds.
+              uint64_t msec;
+              if (az_succeeded(az_span_to_uint64(header.value, &msec))) {
+                retry_after_msec
+                    = msec < INT32_MAX ? (int32_t)msec : INT32_MAX; // int32_t max == ~24 days
+                break;
+              }
+            } else if (az_span_is_equal(header.key, AZ_STR("Retry-After"))) {
+              // The vaule is either seconds or date.
+              uint64_t seconds;
+              if (az_succeeded(az_span_to_uint64(header.value, &seconds))) {
+                retry_after_msec = seconds < INT32_MAX / 1000 ? (int32_t)seconds * 1000 : INT32_MAX;
+                break;
+              }
+              // TODO: Other possible value is HTTP Date. For that, we'll need to parse date, get
+              // current date, subtract one from another, get seconds. And the device should have a
+              // sense of calendar clock.
+            }
+          }
+        } break;
+        default:
+          // None of the retriable HTTP codes above
+          return result;
+      }
+    }
+
+    ++attempt;
+
+    if (retry_after_msec < 0) { // there wasn't any kind of "retry-after" response header
+      int32_t const exponential_retry_after = (int32_t)(
+          (double)retry_delay_msec
+          * (double)(attempt <= 30 ? (1 << attempt) : INT32_MAX) // scale exponentially
+          * (((1.0 / RAND_MAX) * rand()) / 2 + 0.8)); // add jitter factor from 0.8 to 1.3
+
+      retry_after_msec = exponential_retry_after > max_retry_delay_msec ? max_retry_delay_msec
```

Comment:
I also suggest moving all the code that calculates the delay into a separate private function.